### ISSUE TITLE
cadCAD Model todo list!

### ISSUE THREAD

**Issue by GriffGreen** [commons-stack/commons-simulator#45]
## 📋   Description

There are 2 things happening... there is the having a working simulation of a Commons (which will be very useful for the TE Commons), and then there is having a simulation that fits the requirements of the game... there is a lot of overlap but this roadmap is specifically tailored to get cadCAD ready for the game.

For the main Commons Stack Simulation, the goal is to be able to drop in new behavior policies without everything being so messy! This is out of scope for this sprint tho! That is what we do after the sprint but it is important to keep that in the front of your mind when doing the rest of the work.

For the game, the goal is to have a simulation that:
- [ ]  Gives results that make sense and match the proposed behavior of a Commons
- [ ]  Has the exact same output of graphs for any given input of parameters (assuming same randomness seed)
- [x]  Is sensitive to the 6 parameters that the users will be able to tweak (TUNING!)

## 🎉   Definition of done for the first Sprint

- [x]  Onboard a Second dev, Andrew does maintenance and code review, new dev writes code and tests, looks for things that aren't working and carry the project forward.  
- [x]  Defining what a finished cadCAD model looks like for the Simulator!
- [x]  Andrew & Griff agree that this is the path forward
- [x]  Add selling into the simulation
- [x] Create a config file to make it easy to tweak the magic constants that will tune the game's results
- [x] VITO - 1 - (Ask Andrew about this one) Make all policies use the config file to include all the magic constants that live in the simulation code 
- [x] Make a very basic front end that sends in the 6 variables to the cadCAD backend 
- [x]  SANTI - 4 - Update the game.commonsstack.org to work with simrunner.py
- [x]  SANTI/Andrew/Sem - 2 - Display a graph that shows Funding Pool vs time for Main website
- [x]  SANTI/Andrew/Sem - 2 - Display a graph that shows Commons Token Price vs time for Main website
- [x]  SANTI/Andrew/Sem - 2 - Display a graph that shows Sentiment vs time for Main website
- [x]  VITO - 2 - Get the simulation to pump out a variety of reasonable results... Commons that succeed and Commons that fail
- [x]  VITO - 1 - (Ask Andrew about this one) Be able to toggle between randomness or predictable outcomes where any given input always delivers the same output. there is a way to use a seed so that the random values follow a certain pattern

## 🎉   Definition of done for the game

- [x] SANTI - 2 - Tweak the policies of the simulation so that it is sensitive to changes in any one of the 6 variables.
- [x]  Finish the Updating the Proposal Object (below)
- [x]  Finish the Updating the Participant Object (below)
- [x]  VITO - 1 - Finish the Updating the cadCAD Policies (below) 
- [x]  VITO - 2 - Finish the Updating the partial state update blocks (below) (blocked by finishing all the policies

# Definition of finished cadCAD model
## Responsibility
How are responsibilities are divided between the components? The rule is if it is inherent to a particular thing/concept, then that is the responsibility of that class. Otherwise, it belongs to the simulation code, the environment. 

### Proposal
- [x]  is responsible for updating its age
- [x]  is responsible for calculating its conviction threshold
- [x]  is responsible for reporting if it has gathered enough conviction to pass
- [x]  is responsible for reporting if it hasn't enough conviction and will fail

### Participant
- [x]  needs to decide which Proposal to vote for (affinity)
- [x]  needs to decide how much to stake (vote) on each Proposal (rebalance?)
- [x]  needs to decide whether or not to create a new Proposal
- [x]  needs to keep track of his own sentiment
- [x]  Vito - 1 - needs to update his own sentiment if his proposal was passed or failed ( we need to also play with the passing and failing of proposals)
- [x]  decides whether to buy more tokens
- [x]  decides whether to sell tokens
- [x]  decides when to exit (sell all their tokens)

### cadCAD Policies

- [x]  needs to decide how much DAI each participant enters with
- [ ]  ANDREW - 1 - needs to decide which Participant is influential or not (almost done, we just need to use the influence edges)... 
- [x]  ANDREW - 1 - Remove Participant Influence from the simulation  (comment it out) since it is probably computationally intensive and isn't done yet
- [x]  needs to decide which Proposals conflict with each other (not currently used? maybe its worth removing?)
- [x] ANDREW - 1 - Remove Proposal Conflict from the Simulation (comment it out) 
- [x]  needs to decide a new Participant's relationship to the existing network (in network utils)
- [x]  ANDREW - 1 - is responsible for reading from/writing to the graph (The network object where the data is stored... They call the network module...). The changes are indicated by the methods of Participant, Proposal etc (almost done?) 
- [x]  is responsible for simulating income through speculative trades (generate new funding)
- [x]  needs to decide whether an active proposal succeeds or fails
- [x]  needs to calculate overall sentiment (Function is there tho sentiment is not accurate yet)
- [x]  VITO - 1 - New policy: influenced by price + whether proposals are getting completed + Participant's individual sentiment (it calculates the average sentiment... but the sentiment needs to be better spread thru-out the sim)
- [x]  VITO - 1- needs to notify each Participant of Proposal failure/success (it sets the proposal status but it doesn't notify the participants) the Participants should have a change in sentiment here
- [x]  responsible for keeping track of staked tokens and calculating conviction for each Proposal,
- [x] keeps track of token price (obtained from commons object) for eventual plotting
- [x] keeps track of overall sentiment vs time for eventual plotting

## Integration test (complete simulation) goals
- [x] SANTI - 2 - Some simulations should meet the success metrics, whereas some should meet the failure metrics.

## Metrics of Commons Success

- [ ] SANTI - 1 - Change these metrics so that they are reasonable for cadCAD to do
- [ ] SANTI - 2 - If price of the token after 180 days is 30% higher than the hatch price
- [ ] SANTI - 2 - Ratio of successful projects to failed ones is > 50%
- [ ] SANTI - 2 - Overall Sentiment > 0.7????
- [ ] SANTI - 2 - # Participants 180 days after hatch is double the # of Participants at hatch.
- [ ] SANTI - 2 - Over 30 projects were funded in 180 days
- [ ] SANTI - 2 - The Funding Pool has spent 10% more funds than it received at hatch after 700 days 

## Metrics of Commons Failure
- [ ] SANTI - 1 - Change these metrics so that they are reasonable for cadCAD to do
- [ ] SANTI - 2 - If price of the token after 180 days is 20% lower than the hatch price
- [ ] SANTI - 2 - Ratio of successful projects to failed ones is < 30%
- [ ] SANTI - 2 - Overall Sentiment < 0.3
- [ ] SANTI - 2 - # Participants 180 days after hatch is less than the # of Participants at hatch.
- [ ] SANTI - 2 - Under 15 projects were funded in 180 days
- [ ] SANTI - 2 - The Funding Pool has spent only 70% of the funds it received at hatch after 180 days

## Simulation steps
- [x] populate network, setup edges (cadCAD)
- [x] generate new participants (cadCAD)
- [x] generate new proposals (cadCAD)
- [x] generate new funding (cadCAD)
- [x] make active proposals succeed or fail randomly (cadCAD)
- [x] Proposals with enough conviction become active (Proposal)
- [x] decide which candidate Proposals to vote on (Participant)
- [x] rebalance existing + new tokens over all supported Proposals (Participant) 
- [x] decide to buy some more tokens or sell some tokens (Participant)
- [x] and if so how many (bulk buy without slippage may make more sense and be more in line with most Participants' real life experience) (Participant)
- [x] participants may decide exit completely (Participant)

## JSON Schema of data to be sent to frontend

- [x] ANDREW/VITO - 1 - Update the API here... 
```jsonld
{
    "timestep": [1,2,3,4,5...]
    "funding_pool": [funds_timestep_0, funds_timestep_1, funds_timestep_2...]
    "token_supply": [token_supply_timestep_0, token_supply_timestep_1, token_supply_timestep_2...]
    "collateral": [collateral_timestep_0, collateral_timestep_1, collateral_timestep_2...]
    "sentiment": [sentiment_timestep_0, sentiment_timestep_1, sentiment_timestep_2...]
    "token_price": [TO BE DONE]
}
```

## 🤼‍♂️   Reviewer

Griff

**Comment by GriffGreen** [comment]
Closing cause its outdated
